helix-rest/src/main/java/org/apache/helix/rest/metadatastore/ZkMetadataStoreDirectory.java

```diff
@@ -322,14 +322,19 @@ public void refreshRoutingData(String namespace) {
     }
 
     Map<String, List<String>> rawRoutingData;
+    // Remove the raw data first in case of failure on creation
+    _realmToShardingKeysMap.remove(namespace);
```

Comment:
**NOTE**: This check is added for the case when "METADATA_STORE_ROUTING_DATA" is deleted. 
There is one edge case that's *not* handled: when "METADATA_STORE_ROUTING_DATA" contains no node, and then "METADATA_STORE_ROUTING_DATA" gets deleted. In that case, before deletion, the routing data should be empty; after deletion, the routing data should return 404 due to missing namespace. However, since there is no listener on "METADATA_STORE_ROUTING_DATA" itself, its deletion event cannot be handled. Please share you thoughts on whether this edge case should be handled, or if you think this falls into misuse.

Reply:
I don't think you need to worry about the case you're describing. MSD makes the assumption that the root path already exists.